spec/javascripts/components/table-filter-spec.js

```diff
@@ -0,0 +1,100 @@
+/* eslint-env jasmine, jquery */
+/* global GOVUK */
+
+describe('Table Filter component', function () {
+  'use strict'
+
+  var tableFilter
+  var searchInput
+  var element
+
+  beforeEach(function () {
+    var FIXTURE =
+      '<div data-module="table-filter">' +
+        '<div class="govuk-table-filter">' +
+          '<input name="filter">' +
+        '</div>' +
+        '<table>' +
+          '<thead></thead>' +
+          '<tbody class="govuk-table__body">' +
+            '<tr class="govuk-table__row">' +
+              '<td>bcd</td>' +
+              '<td>efg</td>' +
+              '<td>hio</td>' +
+            '</tr>' +
+            '<tr class="govuk-table__row">' +
+              '<td>klm</td>' +
+              '<td>nop</td>' +
+              '<td>£35</td>' +
+            '</tr>' +
+          '</tbody>' +
+        '</table>' +
+        '<p class="govuk-table__message govuk-table__message--hidden">That search returns no results.</p>' +
+      '</div>'
+
+    window.setFixtures(FIXTURE)
+
+    element = document.querySelector('[data-module="table-filter"]')
+    searchInput = document.querySelector('input')
+    tableFilter = new GOVUK.Modules.TableFilter(element)
+  })
+
+  it('Calls the `updateRows` method when the user types in the input field', function () {
+    var event = new window.Event('input')
+
+    spyOn(tableFilter, 'updateRows')
+    tableFilter.init()
+    searchInput.dispatchEvent(event)
+
+    expect(tableFilter.updateRows).toHaveBeenCalledTimes(1)
+  })
+
+  it('Sets the correct classes when the `updateRows` method is called', function () {
+    var rows = document.querySelectorAll('.govuk-table__row')
+    var message = document.querySelector('.govuk-table__message')
+    var hiddenRowClass = tableFilter.hiddenRowClass
+    var hiddenMessageClass = tableFilter.hiddenMessageClass
+
+    // User enters 'b'
+    searchInput.value = 'b'
+    tableFilter.updateRows()
```

Comment:
Why do these tests need to explicitly call the `updateRows` function? If the code has been initialised properly then isn't this unnecessary? I think you should be able to call `tableFilter.init()` once instead of explicitly calling the `updateRows` function multiple times.

Reply:
I think this and the following comment are two sides of the same coin: it is about separating the initialisation of the component from the methods that are called when it has been initialised so that there is not a dependency between them. 

I’ve reduced the number of test cases for this and clarified what they are testing but for now I have retained a call to the method. 

I am aware that this is in effect testing a private method (i.e. one that would not be accessed by a user) but remain to be convinced that this is not viable in this case though I’d be interested in other thoughts on this.

Reply:
I suppose my question would be what is the value of such an approach? There isn't a scenario where separation of the two things you're talking about would happen. You're testing a private method and in doing so you're not testing that the JS as an end-to-end unit works fully i.e. once initialised, user input causes the right effect. I'd prefer to test modules as realistically as possible - this is a small example, but applied to a larger block of code could introduce all kinds of brittleness.

Reply:
I have updated the test to more explicitly check the effects of user input.

Reply:
Great, thanks.